app/src/main/java/com/cp/tasky/auth/shared/data/repository/AuthRepositoryImpl.kt

```diff
@@ -5,30 +5,50 @@ import com.cp.tasky.auth.shared.data.mapper.LoginUserMapper
 import com.cp.tasky.auth.shared.domain.AuthRepository
 import com.cp.tasky.auth.shared.domain.model.LoginResponse
 import com.cp.tasky.auth.shared.domain.model.UserCredential
+import com.cp.tasky.core.data.util.NetworkErrorMessage
 import com.cp.tasky.core.data.util.NetworkResult
 import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
 
 class AuthRepositoryImpl(
     private val remoteDataSource: RemoteDataSource
 ) : AuthRepository {
 
     override suspend fun loginUser(userCredential: UserCredential): NetworkResult<LoginResponse> {
         return try {
-            val response = remoteDataSource.loginUser(
-                LoginApiBody(
-                    email = userCredential.email, password = userCredential.password
+            withContext(Dispatchers.IO) {
+                val response = remoteDataSource.loginUser(
+                    LoginApiBody(
+                        email = userCredential.email, password = userCredential.password
+                    )
                 )
-            )
 
-            NetworkResult.Success(
-                LoginUserMapper.toLoginResponse(response)
-            )
+                if (response.isSuccessful) {
+                    response.body()?.let {
+                        NetworkResult.Success(
+                            LoginUserMapper.toLoginResponse(it)
+                        )
+                    } ?: run {
+                        NetworkResult.Error(
+                            message = NetworkErrorMessage.SERVER_NOT_AVAILABLE
+                        )
+                    }
+                } else {
+                    when (response.code()) {
+                        in 400..499 -> NetworkResult.Error(message = NetworkErrorMessage.INVALID_EMAIL_OR_PASSWORD)
```

Comment:
This is too generalized. It's only an invalid email/password if you get 401 Unauthorized back. There are lots of different status codes that all have a different meaning